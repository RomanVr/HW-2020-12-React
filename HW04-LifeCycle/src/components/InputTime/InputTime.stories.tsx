import React from "react";
import { text, withKnobs } from "@storybook/addon-knobs";
import { InputTime } from "..";

export default {
  title: "InputTime",
  Component: InputTime,
  decorators: [withKnobs],
};

export const InputTimeStory: React.FC = () => {
  return <InputTime timeValue={text("filled with", "12:05:55")} />;
};
