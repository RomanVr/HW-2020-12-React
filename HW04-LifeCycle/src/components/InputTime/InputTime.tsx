import React from "react";

interface InputTimeProps {
  timeValue: string;
}

interface InputTimeState {
  timeValue: string;
}

export class InputTime extends React.Component<InputTimeProps, InputTimeState> {
  // eslint-disable-next-line no-undef
  timerID?: NodeJS.Timeout;

  constructor(props: InputTimeProps) {
    super(props);
    this.state = { timeValue: new Date().toLocaleTimeString() };
  }

  componentDidMount(): void {
    this.timerID = setInterval(() => this.tick(), 1000);
  }

  componentWillUnmount(): void {
    if (this.timerID) {
      clearInterval(this.timerID);
      this.timerID = undefined;
    }
  }

  tick(): void {
    this.setState({
      timeValue: new Date().toLocaleTimeString(),
    });
  }

  render(): React.ReactElement {
    return <input readOnly type="text" value={this.state.timeValue} />;
  }
}
