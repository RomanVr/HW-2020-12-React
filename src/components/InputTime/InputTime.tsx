import React from "react";

interface InputTimeState {
  timeValue: number;
}

export class InputTime extends React.Component<unknown, InputTimeState> {
  // eslint-disable-next-line no-undef
  timerID?: NodeJS.Timeout;

  constructor(props: never) {
    super(props);
    this.state = { timeValue: 0 };
  }

  moment(timeSeconds: number): string {
    function num(val: number) {
      val = Math.floor(val);
      return val < 10 ? "0" + val : val;
    }
    const hours = (timeSeconds / 3600) % 24;
    const minutes = (timeSeconds / 60) % 60;
    const seconds = timeSeconds % 60;
    return `${num(hours)}:${num(minutes)}:${num(seconds)}`;
  }

  componentDidMount(): void {
    this.timerID = setInterval(() => this.tick(), 1000);
  }

  componentWillUnmount(): void {
    clearInterval(Number(this.timerID));
    this.timerID = undefined;
  }

  tick(): void {
    this.setState({
      timeValue: this.state.timeValue + 1,
    });
  }

  render(): React.ReactElement {
    const timeValue = this.moment(this.state.timeValue);
    return (
      <input
        data-testid="inputTime"
        readOnly
        type="text"
        value={timeValue}
        style={{ width: timeValue.length * 7, textAlign: "center" }}
      />
    );
  }
}
