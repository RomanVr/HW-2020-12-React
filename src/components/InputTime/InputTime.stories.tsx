import React from "react";
import { Meta, Story } from "@storybook/react";

import { InputTime } from "./InputTime";

export default {
  title: "Form/InputTime",
  component: InputTime,
} as unknown as Meta;

export const InputTimeStory: Story = () => <InputTime />;
