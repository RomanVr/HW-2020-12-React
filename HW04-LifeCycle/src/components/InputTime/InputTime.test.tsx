import React from "react";
import renderer from "react-test-renderer";
import { InputTime } from "./InputTime";

describe("InputTime", () => {
  it("Render input for with value Time", () => {
    expect(
      renderer.create(<InputTime timeValue="Time" />).toJSON()
    ).toMatchSnapshot();
  });
});
