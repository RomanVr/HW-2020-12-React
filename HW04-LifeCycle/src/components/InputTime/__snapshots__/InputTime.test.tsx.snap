// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTime Render input for with value Time 1`] = `
<input
  readOnly={true}
  type="text"
  value="22:56:00"
/>
`;
